modules/earning/src/lib.rs

```diff
@@ -0,0 +1,259 @@
+
+	#[pallet::pallet]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(T::WeightInfo::bond())]
+		#[transactional]
+		pub fn bond(origin: OriginFor<T>, #[pallet::compact] amount: Balance) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			let change = <Self as BondingController>::bond(&who, amount)?;
+
+			if let Some(change) = change {
+				T::OnBonded::happened(&(who.clone(), change.change));
+				Self::deposit_event(Event::Bonded {
+					who,
+					amount: change.change,
+				});
+			}
+			Ok(())
+		}
+
+		#[pallet::weight(T::WeightInfo::unbond())]
+		#[transactional]
+		pub fn unbond(origin: OriginFor<T>, #[pallet::compact] amount: Balance) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			let unbond_at = frame_system::Pallet::<T>::block_number().saturating_add(T::UnbondingPeriod::get());
+			let change = <Self as BondingController>::unbond(&who, amount, unbond_at)?;
+
+			if let Some(change) = change {
+				T::OnUnbonded::happened(&(who.clone(), change.change));
+				Self::deposit_event(Event::Unbonded {
+					who,
+					amount: change.change,
+				});
+			}
+
+			Ok(())
+		}
+
+		#[pallet::weight(T::WeightInfo::unbond())]
+		#[transactional]
+		pub fn unbond_instant(origin: OriginFor<T>, #[pallet::compact] amount: Balance) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			let change = <Self as BondingController>::unbond_instant(&who, amount)?;
+
+			if let Some(change) = change {
+				let amount = change.change;
+				let fee = T::InstantUnstakeFee::get().mul_ceil(amount);
+				let final_amount = amount.saturating_sub(fee);
+
+				let unbalance =
+					T::Currency::withdraw(&who, fee, WithdrawReasons::TRANSFER, ExistenceRequirement::KeepAlive)?;
+				T::OnUnstakeFee::on_unbalanced(unbalance);
+
+				T::OnUnbonded::happened(&(who.clone(), final_amount));
+				Self::deposit_event(Event::InstantUnbonded {
+					who,
+					amount: final_amount,
+					fee,
+				});
+			}
+
+			Ok(())
+		}
+
+		#[pallet::weight(T::WeightInfo::rebond())]
+		#[transactional]
+		pub fn rebond(origin: OriginFor<T>, #[pallet::compact] amount: Balance) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			let change = <Self as BondingController>::rebond(&who, amount)?;
+
+			if let Some(change) = change {
+				T::OnBonded::happened(&(who.clone(), change.change));
+				Self::deposit_event(Event::Rebonded {
+					who,
+					amount: change.change,
+				});
+			}
+
+			Ok(())
+		}
+
+		#[pallet::weight(T::WeightInfo::withdraw_unbonded())]
+		#[transactional]
+		pub fn withdraw_unbonded(origin: OriginFor<T>) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			let change =
+				<Self as BondingController>::withdraw_unbonded(&who, frame_system::Pallet::<T>::block_number())?;
```

Comment:
having `withdraw_unbonded` return Ok(None) when no funds are available to withdraw is more in line with the other methods in BondController